src/eckit/stats/method/MethodT.cc

```diff
@@ -0,0 +1,122 @@
+/*
+ * (C) Copyright 1996- ECMWF.
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * In applying this licence, ECMWF does not waive the privileges and immunities
+ * granted to it by virtue of its status as an intergovernmental organisation nor
+ * does it submit to any jurisdiction.
+ */
+
+
+#include "eckit/stats/method/MethodT.h"
+
+#include <cmath>
+
+#include "eckit/stats/Field.h"
+#include "eckit/stats/detail/AngleT.h"
+#include "eckit/stats/detail/CentralMomentsT.h"
+#include "eckit/stats/detail/ScalarT.h"
+
+
+namespace eckit::stats::method {
+
+
+template <typename STATS>
+void MethodT<STATS>::resize(size_t N) {
+    vector_t::resize(N);
+}
+
+
+template <typename STATS>
+void MethodT<STATS>::execute(const Field& field) {
+    ASSERT(!empty());
+    Counter::reset(field);
+
+    ASSERT(field.dimensions() == 1);
+    ASSERT(field.values(0).size() == size());
+
+    auto v = field.values(0).begin();
+    for (auto& s : *this) {
+        const auto& value = *(v++);
+        if (Counter::count(value)) {
+            s(value);
+        }
+    }
+
+    ASSERT(v == field.values(0).end());
+}
+
+
+template <typename STATS>
+void MethodT<STATS>::mean(Field& field) const {
+    const auto missingValue = field.missingValue();
+
+    ASSERT(field.dimensions() == 1);
+    ASSERT(field.values(0).size() == size());
+
+    MIRValuesVector statistics(field.values(0).size());
+    auto v = statistics.begin();
+
+    for (auto& s : *this) {
+        auto& value = *(v++);
+        auto stat   = s.mean();
+        value       = std::isnan(stat) == 0 ? stat : missingValue;
+    }
+
+    field.update(statistics, 0, true);
+}
+
+
+template <typename STATS>
+void MethodT<STATS>::variance(Field& field) const {
+    const auto missingValue = field.missingValue();
+
+    ASSERT(field.dimensions() == 1);
+    ASSERT(field.values(0).size() == size());
+
+    MIRValuesVector statistics(field.values(0).size());
+    auto v = statistics.begin();
+
+    for (auto& s : *this) {
+        auto stat = s.variance();
+        *v        = std::isnan(stat) == 0 ? stat : missingValue;
+    }
+
+    field.update(statistics, 0, true);
+}
+
+
+template <typename STATS>
+void MethodT<STATS>::stddev(Field& field) const {
+    const auto missingValue = field.missingValue();
+
+    ASSERT(field.dimensions() == 1);
+    ASSERT(field.values(0).size() == size());
+
+    MIRValuesVector statistics(field.values(0).size());
+    auto v = statistics.begin();
+
+    for (auto& s : *this) {
+        auto stat = s.standardDeviation();
+        *v        = std::isnan(stat) == 0 ? stat : missingValue;
+    }
+
+    field.update(statistics, 0, true);
+}
+
+
+static const MethodBuilder<MethodT<detail::AngleT<double, detail::AngleScale::DEGREE, detail::AngleSpace::ASYMMETRIC>>>
+    __stats1("angle.degree.asymmetric");
```

Comment:
Not sure I like the variable names starting with __. Those are supposed to be reserved symbols.

If you really want to ensure non-clashing, you can put them in an anonymous namespace?

Reply:
Pushed